[visionOS] Optimize the QuickLook fullscreen transition

etiennesegonzac opened the pull request:
#### 70014c526e8ece8ced6099c575ac1ec32d75f353
<pre>
[visionOS] Optimize the QuickLook fullscreen transition
<a href="https://bugs.webkit.org/show_bug.cgi?id=275686">https://bugs.webkit.org/show_bug.cgi?id=275686</a>
&lt;<a href="https://rdar.apple.com/129195533">rdar://129195533</a>&gt;

Reviewed by Mike Wyrzykowski.

Scene swapping takes care of the transition when using QuickLook
fullscreen. In order to improve performance and avoid memory spikes due
to DynamicContentScalingDisplayList generation, introduce a new
fullscreen code path where we freeze the layer tree and skip the
expensive WebView snapshoting/reparenting.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController exitFullScreen]):
Introduce new QuickLook fullscreen branches with early returns.
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Remove the old QuickLook fullscreen branches.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
Introduce a new `LayerTreeFreezeReason`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::didExitFullScreen):
Freeze the layer tree while using QuickLook fullscreen.

Canonical link: <a href="https://commits.webkit.org/280896@main">https://commits.webkit.org/280896@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4ba2f1435333719d545b2953e2cb270ea9892f5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61378 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8201 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46795 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5819 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49909 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27620 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31548 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7205 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53500 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7472 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63061 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1670 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54015 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1676 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49920 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54133 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12821 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1415 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32913 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33999 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35083 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33744 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -875,15 +875,44 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
+            [_previewWindowController presentWindow];
+            _fullScreenState = WebKit::InFullScreen;
+
+            if (auto* manager = strongSelf.get()._manager) {
```
```suggestion
            if (auto* manager = [strongSelf _manager]) {
```

pxlcoder commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -441,6 +446,12 @@ static Vector<Ref<Element>> collectFullscreenElementsFromElement(Element* elemen
 void WebFullScreenManager::didExitFullScreen()
 {
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
+    if (m_willUseQuickLookForFullscreen)
```
```suggestion
    if (std::exchange(m_willUseQuickLookForFullscreen, false))
```

pxlcoder commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -267,6 +269,9 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
     }
+
+    if (m_willUseQuickLookForFullscreen)
+        m_page->freezeLayerTree(WebPage::LayerTreeFreezeReason::PageTransition);
```
We may want a new, specific `LayerTreeFreezeReason` to avoid it getting reset elsewhere.

pxlcoder reviewed (commented):
Also flagging Simon and Jer for review, to comment on whether freezing the layer tree is the right thing to do here.

jernoble commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -882,15 +882,44 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
         return;
 
     OBJC_ALWAYS_LOG(OBJC_LOGIDENTIFIER, WebCore::FloatSize { mediaDimensions });
+    __block auto logIdentifier = OBJC_LOGIDENTIFIER;
```
This shouldn't need to be marked as a `__block` variable, since you're using it in a C++ lambda; in fact, this could just be defined inside the capture declaration....

jernoble commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -882,15 +882,44 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
+        OBJC_ALWAYS_LOG(logIdentifier, "(QL) presentation updated");
+        manager->willEnterFullScreen();
+
+        manager->prepareQuickLookImageURL([strongSelf = retainPtr(self), self, window = retainPtr([webView window]), logIdentifier = logIdentifier] (URL&& url) mutable {
```
...here:

```
        manager->prepareQuickLookImageURL([strongSelf = retainPtr(self), self, window = retainPtr([webView window]), logIdentifier = OBJC_LOGIDENTIFIER] (URL&& url) mutable {
```

jernoble commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -882,15 +882,44 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
+        _imageDimensions = mediaDimensions;
+        _fullScreenState = WebKit::WaitingToEnterFullScreen;
+
+        OBJC_ALWAYS_LOG(logIdentifier, "(QL) presentation updated");
```
(out-of-order comment: best practice is to have at most one "always log" per method, taking early returns into account. This would mean moving the "always log" on line 884 below the `if (_isUsingQuickLook)` check.)

jernoble commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -882,15 +882,44 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
+            _fullScreenState = WebKit::InFullScreen;
+
+            if (auto* manager = [strongSelf _manager]) {
+                OBJC_ALWAYS_LOG(logIdentifier, "(QL) presentation completed");
```
"Always logging" in completion handlers is perfectly fine though!

jernoble commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -977,7 +1001,6 @@ - (void)enterFullScreen:(CGSize)mediaDimensions
     [[_webViewPlaceholder layer] setName:@"Fullscreen Placeholder View"];
 
     WKSnapshotConfiguration* config = nil;
-    __block auto logIdentifier = OBJC_LOGIDENTIFIER;
```
That would mean this line could stay here.

mwyrzykowski reviewed (commented):
This looks good to me but it would be good to get input from @pxlcoder or @jernoble

etiennesegonzac committed:
[visionOS] Optimize the QuickLook fullscreen transition
https://bugs.webkit.org/show_bug.cgi?id=275686
<rdar://129195533>

Reviewed by Mike Wyrzykowski.

Scene swapping takes care of the transition when using QuickLook
fullscreen. In order to improve performance and avoid memory spikes due
to DynamicContentScalingDisplayList generation, introduce a new
fullscreen code path where we freeze the layer tree and skip the
expensive WebView snapshoting/reparenting.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController exitFullScreen]):
Introduce new QuickLook fullscreen branches with early returns.
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Remove the old QuickLook fullscreen branches.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
Introduce a new `LayerTreeFreezeReason`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::didExitFullScreen):
Freeze the layer tree while using QuickLook fullscreen.

Canonical link: https://commits.webkit.org/280896@main
